src/main/java/org/ossgang/commons/monads/AsyncMaybe.java

```diff
@@ -153,17 +141,14 @@ protected AsyncMaybe(CompletableFuture<T> future, Function<ThrowingSupplier<T>,
      * @return a new {@link AsyncMaybe} with the same result as this {@link AsyncMaybe} or with an exception if one thrown in the provided consumer
      */
     public AsyncMaybe<T> whenComplete(ThrowingConsumer<Maybe<T>> consumer) {
-        return AsyncMaybe.fromCompletableFuture(future.whenCompleteAsync((value, exception) -> {
-            if (exception != null) {
-                uncheckedConsumer(consumer).accept(Maybe.ofException(exception));
-            } else {
-                uncheckedConsumer(consumer).accept(maybeGenerator.apply(() -> value));
-            }
-        }, ASYNC_MAYBE_POOL));
+        return new AsyncMaybe<>(stage.thenApplyAsync(stageResult ->
+                Maybe.attempt(() -> consumer.accept(stageResult)).flatMap(any -> stageResult), ASYNC_MAYBE_POOL));
```

Comment:
Why do you wrap the consumer in Maybe.attempt() here but not in the other whenXxx methods?

Reply:
Here we need to pass the entire `stageResult` to the consumer and then return a maybe that is an error if the consumer throws or the `stageResult` if it's ok.. I figured the attempt + flatMap was quite elegant. The others are easily solvable with `.ifValue/Exception` since their consumer only accepts one of them :slightly_smiling_face:

Reply:
Yes, but be careful - ifValue and ifException do not catch exceptions of the consumer (hence they don't accept ThrowingConsumer), they just return `this` for chaining.

So in this case, we wrap exceptions thrown by the consumer, in the other cases we let them escape to the CompleteableFuture.

See https://github.com/ossgang/ossgang-commons/blob/master/src/main/java/org/ossgang/commons/monads/Maybe.java#L369

Reply:
Not good ! :innocent: let's discuss offline then :wink:

Reply:
Discussed offline and now with `Maybe#apply` and `Maybe#flatApply` the solution works